### DISCUSSION TITLE
Error accessing Entity element using .Net Core 3.1

### DISCUSSION THREAD

**Discussion by Rick6561** [pythonnet/pythonnet#1960] [Q&A]
Using Python 3.8.10
Pythonnet 3.0

Eventually when calling into a .netCore all that I wrote that creates a context object the following exception is generated:

The type initializer for 'NSIMI.Data.BaseContext`1' threw an exception. ---> System.TypeLoadException: Could not load type 'System.Data.Common.DbProviderFactories' from assembly 'netstandard, Version=2.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'.
   at System.Data.Entity.Infrastructure.DependencyResolution.DefaultProviderFactoryResolver.GetService(Type type, Object key, Func`3 handleFailedLookup)
   at System.Data.Entity.Infrastructure.DependencyResolution.DefaultProviderFactoryResolver.GetServices(Type type, Object key)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at System.Linq.Enumerable.<SelectManyIterator>d__17`2.MoveNext()
   at System.Linq.Enumerable.<ConcatIterator>d__59`1.MoveNext()
   at System.Linq.Enumerable.<SelectManyIterator>d__17`2.MoveNext()
   at System.Linq.Enumerable.<ConcatIterator>d__59`1.MoveNext()
   at System.Linq.Enumerable.<OfTypeIterator>d__95`1.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at System.Data.Entity.Infrastructure.DependencyResolution.InternalConfiguration.Lock()
   at System.Data.Entity.Infrastructure.DependencyResolution.DbConfigurationManager.<.ctor>b__10_0()
   at System.Lazy`1.CreateValue()
   at System.Lazy`1.LazyInitValue()
   at System.Data.Entity.Infrastructure.DependencyResolution.DbConfigurationManager.GetConfiguration()
   at System.Data.Entity.Database.SetInitializer[TContext](IDatabaseInitializer`1 strategy)
   at NSIMI.Data.BaseContext`1..cctor()

It is my understanding that System.Data.Common.DbProviderFactories is not supported in netstandard 2.0. 

While writing this I went to look at the version of pythonnet and I opened the runtime.deps.json file. In it it has .NETStandard,Version=v2.0" as a target. So perhaps this explains my problem.

Can this get targeted for 2.1? Can I work this out or is pythonnet not ready for .NETStandard,Version=v2.1" ?

Thx
Rick

**Comment by Rick6561** [comment]
Follw-up... I downloaded source and recompiled Python.Runtime changing this line: <TargetFrameworks>netstandard2.0</TargetFrameworks>
to
<TargetFrameworks>netstandard2.1</TargetFrameworks>

Compiles but now 
import clr generates an exception: 
  Message=cannot call null pointer pointer from cdata 'int(*)(void *, int)'
  
Forgot to mention API that I wrote that this python code is trying to call into is targeted for Core 3.1

**Reply by filmor** [reply]
Whether Python.NET is .NET Standard 2.0 or 2.1 should be irrelevant and its runtimeconfig is never loaded. Just make sure that your application code is properly `dotnet publish` and use its runtime config.

**Comment by Rick6561** [comment]
Ok. The runtime I am using:
rt = get_coreclr("C:\\Users\\XXX\\source\\repos\\XXX\\XXX\\Main\\bin\\Debug\\netcoreapp3.1\\ApplicationInterface.runtimeconfig.json")

It looks like:
{
  "runtimeOptions": {
    "tfm": "netcoreapp3.1",
    "framework": {
      "name": "Microsoft.WindowsDesktop.App",
      "version": "3.1.0"
    }
  }
}

Is there more needed in the runtime?

**Reply by filmor** [reply]
I never tried to use the desktop runtime, I have no idea whether that one will work.

**Comment by Rick6561** [comment]
I have library now that focuses entirely on the offending operation, creating a DbContext

using System.Data.Entity;

namespace XXX.APIRedux
{
    public class ApplicationInterface
    {
        public ApplicationInterface()
        {
        }

        public static void LaunchAcquisitionEnginePython(int acquisDefID)
        {
            SchoolDBEntities model = new SchoolDBEntities();
        }
    }

    public partial class SchoolDBEntities : DbContext
    {
        public SchoolDBEntities()
            : base("name=SchoolDBEntities")
        {
        }
    }
}


Python Code at its simplest:
from clr_loader import get_coreclr
from pythonnet import set_runtime
import os
import clr

print ("start api")

json = (r"C:\xxx\APIRedux.runtimeconfig.json")

rt = get_coreclr(json)

set_runtime(rt)

clr.AddReference(r"xxx\APIRedux.dll")

from XXX.APIRedux import ApplicationInterface

print(ApplicationInterface.LaunchAcquisitionEnginePython(147))

Referenced runtimeConfig not looks like this:
{
  "runtimeOptions": {
    "tfm": "netcoreapp3.1",
    "framework": {
      "name": "Microsoft.NETCore.App",
      "version": "3.1.0"
    }
  }
}

And I am still getting this exception:

The type initializer for 'XXX.Data.BaseContext1' threw an exception. ---> System.TypeLoadException: Could not load type 'System.Data.Common.DbProviderFactories' from assembly 'netstandard, Version=2.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'. at System.Data.Entity.Infrastructure.DependencyResolution.DefaultProviderFactoryResolver.GetService(Type type, Object key, Func3 handleFailedLookup)

I don't know how to get it off of the netstandard 2.0.0.0

What am I missing?

Thx
Rick

**Comment by Rick6561** [comment]
You said above the 2.0.0.0 shouldn't matter but I know the factory is not supported in 2.0.0.0

**Reply by Rick6561** [reply]
EntityFramework (System.Data.Entity) depends on .NetStandard, Version v2.1

**Reply by filmor** [reply]
It does not matter. I don't know what else to say. The .NET Standard version of Python.Runtime just defines a set of minimal versions of supported .NET runtimes, not the maximum.
